Recursive inlining via InlineTransform and associated fixes

mlange05 opened the pull request:
This PR separates the various styles of source-inlining (elementals, internals, pragma-marked) into a dedicated `Transformation` class that can be configured independently. This key motivation for this is that this allows us to perform recursive multi-step inlining, where individual inlining steps are performed at the leaves first and then propagated up by traversing the `Scheduler` tree in reverse order.

As these steps where previously baked into the `SCCBaseTransformation` several of these steps have now been removed. Due to complex interplay in the ec-physics regression, things like associate resolution and dead-code-elimination have to now also be performed with the inlining steps. Until we have proper pipeline configuration, I have chosen to go the simple route and simply lump these in with the `InlineTransformation`, but would appreciate feedback on this. This refactoring also has show some (severe!) problems with the previous "elemental function" inliner, which are also fixed in this PR. 

Edit: As discussed offline, I've also added a first draft of a "housekeeping" transformation (`SanitiseTransformation`), for a easily configurable code harmonisation utility pass. I plan to extend this as we progress through the trafor cleanup and consolidation.

Edit: Turns out one of the small fixes actually has a some far-reaching consequences: The elemental inliner was originally not checking for true elemental functions, and it now does(!), but the problem is that our OFP frontend cannot deal with prefix keywords. As the elemental inliner is needed for C-transpilation though, this means a bunch of tests, including CLOUDSC, have to be disabled for OFP. I've had a go at retro-fitting this feature, but it's quite involed due to the broken scoping around type nodes in the OFP AST. I propose dropping OFP entirely now, but I'll track that as a separate issue.

reuterbal commented on loki/transform/transform_inline.py:
```diff
@@ -193,7 +285,7 @@ def inline_elemental_functions(routine):
     import_map = {}
     for im in FindNodes(Import).visit(routine.spec):
-        if all(hasattr(s, 'type') and s.type.dtype in removed_functions for s in im.symbols):
+        if im.symbols and all(hasattr(s, 'type') and s.type.dtype in removed_functions for s in im.symbols):
```
I think the `hasattr(s, 'type')` should no longer be necessary - that is a remnant of the times when we represented the symbol list in `Import` with names as strings.

reuterbal commented on transformations/transformations/single_column_coalesced.py:
```diff
@@ -308,27 +288,10 @@ def process_kernel(self, routine):
-
         # Associates at the highest level, so they don't interfere
         # with the sections we need to do for detecting subroutine calls
         resolve_associates(routine)
```
What's the reason for not including `resolve_associates` in the `SanitiseTransformation`?

mlange05 replied:
Oversight, now fixed (I hope).

reuterbal replied:
Sure, I only asked for the reason, and that's a good enough one ;-)

reuterbal reviewed (changes requested):
This looks great and is much appreciated!

I left a few comments while browsing the changes, most of which are not strictly necessary to act upon other than some massaging of docstrings (marking this as Request changes for that). One thing to consider, though, would be to add a test for the new `SanitiseTransformation` to the Loki core's test base. It's currently only being invoked from the tests in the transformations sub-package.

mlange05 commented on loki/transform/transform_inline.py:
```diff
@@ -179,7 +268,10 @@ def inline_elemental_functions(routine):
     exprmap = {}
     for call in FindInlineCalls().visit(routine.body):
-        if call.procedure_type is not BasicType.DEFERRED:
+        if call.procedure_type is BasicType.DEFERRED:
```
Yes, I'm planning an overhaul of the logging from transformations once the configuration changes are more complete (so we can select log-leve per trafo, etc.).

mlange05 replied:
Possibly worth capturing via an issue?

reuterbal replied:
#220

mlange05 commented on loki/transform/transform_inline.py:
```diff
@@ -32,10 +35,88 @@
+        Strip unused import symbols after pragma-inlining (optional, default: True)
+    external_only : bool, optional
+        Do not replace variables declared in the local scope when
+        inlining constants (default: True)
```
Yes!

mlange05 commented:
Ok, I rebased and pushed the fixes to the suggestions - many thanks @reuterbal . I'll have to rebase/re-run this with ec-phys regression, just to make sure, but will ping here once confirmed.

mlange05 commented:
Ok, rebased, CI happy and confirmed with EC-physics regression. I've also added test for the new trafo and addressed all comments, except the breakage one. I think this is ready for another look, @reuterbal

reuterbal reviewed (approved):
Many thanks, changes look good to me and everything has been addressed. GTG! :shipit:
